### PR TITLE
Develop Shop Review Management Feature

### PR DESCRIPTION
**Pull request by samcyn**
## Description:
This pull request implements CRUD (Create, Read, Update, Delete) functionality for managing reviews for a given shop. It covers the following operations:

- Create: Allows users to add a new review for a shop.
- Read: Enables fetching all reviews for a shop or retrieving a specific review by its ID.
- Update: Permits users to edit the content or rating of an existing review.
- Delete: Allows users to remove a review.

## Changes Made:
- Implemented REST API endpoints for CRUD operations related to shop reviews.
- Added models and schema for storing shop reviews.
- Integrated validation and error handling mechanisms for input data.
- Updated frontend components to interact with the new API endpoints.

## Rationale:
The goal of this PR is to provide users with the ability to manage reviews for shops within the application. By implementing CRUD functionality, users can seamlessly add, view, edit, and delete reviews, enhancing the overall user experience and engagement with the platform.

## Approach:
- Utilized Vue3 and tailwind css for development, leveraging on Mockapi.io.
- Set up eslint, prettier and commitlint to keep up with best practices.
- A lean and straightforward directory set up 
- Model definition of Product, Shop and Review and their relationships.
- Followed RESTful API design principles for defining routes and endpoints.
- Set up firebase for authentication
- Employed Route guards for authentication and authorization to ensure secure access to review management functionality.
- Global error handling
- Form validations using Veevalidate
- Implemented unit tests using Vitest for robustness and reliability.
- Caching of data to manage numbers of network requests
- Pagination, sorting, searching and filtering,
- Responsive and mobile first approach.

## Testing:
- Developed comprehensive unit tests to verify the correctness of CRUD operations and edge cases.
- Tested API endpoints using Postman to ensure functionality and compatibility.

## Dependencies:
This PR depends on the following:
- Vue3
- Tailwind CSS
- firebase for Authentication
- Axios
- Veevalidate

## Screenshots/Videos:
[Demo Video](https://www.loom.com/share/823c0292498649379e1d9413308f2dce)

## Instructions to Run the App:
1. Clone the repository 
```bash
git clone https://github.com/judgeme/recruitment.git
cd recruitment
```
2. Install all dependency
```bash
yarn install
```
3. Set up environment variables
- Create a .env file in the root directory.
- Add your Firebase configuration
```bash
VITE_FIREBASE_API_KEY =
VITE_FIREBASE_AUTH_DOMAIN =
VITE_FIREBASE_PROJECT_ID =
VITE_FIREBASE_STORAGE_BUCKET =
VITE_FIREBASE_MESSAGING_SENDER_ID =
VITE_FIREBASE_APP_ID =
```
4. Run the development server
```bash
yarn dev
```
5. Open your browser and navigate to:
```bash
http://localhost:5173/
```

## Live Deployment:
The live deployment of the application can be accessed [Live Demo](https://judgeme-gamma.vercel.app).

## Test Account:
To test the functionality, you can use the following test account:

- Email: tester@samsoniyanda.com
- Password: somerandomtext101@

Feel free to log in and explore the review management features.

### CODE REVIEW

File: tasks/frontEnd/.eslintrc.cjs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    {
+      files: ['e2e/**/*.{test,spec}.{js,ts,jsx,tsx}'],
+      extends: ['plugin:playwright/recommended']
+    }
+  ],
+  parserOptions: {
+    ecmaVersion: 'latest'
+  },
+  rules: {
+    'vue/multi-word-component-names': 'off',
```
Comment (MarcusAl) [inline tasks/frontEnd/.eslintrc.cjs:22]: Any reason for overriding this rule?

File: tasks/frontEnd/.eslintrc.cjs [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    semi: ['error', 'always']
```
Comment (MarcusAl) [inline tasks/frontEnd/.eslintrc.cjs:23]: Any particular reason for enforcing this?

File: tasks/frontEnd/.github/workflows/pull-request-lint.yml [hunk 3]
```diff
@@ -0,0 +1,34 @@
+name: CI
```
Comment (MarcusAl) [inline tasks/frontEnd/.github/workflows/pull-request-lint.yml:1]: Nit: I would usually use the whole word or camelcase for folder naming, e.g. `/frontend` or `/front_end`

File: tasks/frontEnd/.vscode/extensions.json [hunk 4]
```diff
@@ -0,0 +1,8 @@
+{
+  "recommendations": [
+    "Vue.volar",
+    "ms-playwright.playwright",
```
Comment (MarcusAl) [inline tasks/frontEnd/.vscode/extensions.json:4]: Why use playwright instead of vitest?

File: tasks/frontEnd/e2e/auth-flow-login.spec.ts [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+  // Mock the api call - Login
+  await page.route('https://identitytoolkit.googleapis.com/v1/accounts:signInWithPassword?key=*', async route => {
+    await route.fulfill({ json: MOCKED_LOGIN_RESPONSE });
+  });
+
+  // Mock the api call - Account Look Up
+  await page.route('https://identitytoolkit.googleapis.com/v1/accounts:lookup?key=*', async route => {
+    await route.fulfill({ json: MOCKED_ACCOUNT_LOOK_UP_RESPONSE });
+  });
+
```
Comment (MarcusAl) [inline tasks/frontEnd/e2e/auth-flow-login.spec.ts:14]: Nice to see tests, but perhaps these can be separated into concerns, and positive/negative scenarios?

E.g. What happens when the response returns an error. What is happening in each separate stage?

File: tasks/frontEnd/e2e/auth-flow-login.spec.ts [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+  await page.getByRole('button', { name: 'log out' }).click();
+  await expect(page).toHaveURL('http://localhost:5173/login?redirect=/');
+  await page.getByPlaceholder('Email').click();
+  await page.getByPlaceholder('Email').fill('tester@samsoniyanda.com');
+  await page.getByPlaceholder('Password').click();
+  await page.getByPlaceholder('Password').fill('document007');
+  await page.getByPlaceholder('Password').press('CapsLock');
+  await page.getByPlaceholder('Password').fill('somerandomtext');
+  await page.getByRole('button', { name: 'Login' }).click();
+  await expect(page).toHaveURL('http://localhost:5173');
```
Comment (MarcusAl) [inline tasks/frontEnd/e2e/auth-flow-login.spec.ts:48]: Nit: `http://localhost:5173` could be stored in a constant, so that we can flexibly set the value in different environments

File: tasks/frontEnd/src/components/AppHeader/index.vue [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+              {{ route.name === 'login' ? 'Sign Up' : 'Login' }}
+            </app-link>
+          </li>
+        </ul>
+        <button aria-label="menu" type="button" class="md:hidden" @click="toggleMenu">
+          <app-icon icon="hamburger" :size="40" viewBox="0 0 512 512" />
+        </button>
+      </nav>
+    </div>
+    <!-- mobile menu -->
```
Comment (MarcusAl) [inline tasks/frontEnd/src/components/AppHeader/index.vue:188]: 👍

File: tasks/frontEnd/.eslintrc.cjs (hunk 1 above)
Comment (samcyn) [reply tasks/frontEnd/.eslintrc.cjs:22]: Yes! Disabling that setting allows us to name the Vue component file as index.vue. This is beneficial for avoiding multiple import statements. For instance, in the AppButton folder, I want to include tests, a Storybook story, and the Vue component itself. Without turning off this rule, I would be forced to import components like this: ../AppButton/MyButtonComponent.vue instead of the cleaner ../AppButton/index.vue.
This approach enhances code clarity, especially when dealing with deeply nested component structures. However, it's important to note that when using this method, it's best practice to specify the component name within the defineOptions API, as demonstrated.

File: tasks/frontEnd/src/components/shared/AppActivityLoader/AppActivityLoader.test.ts [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+      slots: {
+        default: '<p>Some default content here</p>',
+      },
+    });
+
+    // Check if the component is rendered
+    expect(wrapper.exists()).toBe(true);
+
+    // Check if the AppLoader is not rendered
+    expect(wrapper.findComponent(AppLoader).exists()).toBe(false);
```
Comment (MarcusAl) [inline tasks/frontEnd/src/components/shared/AppActivityLoader/AppActivityLoader.test.ts:19]: Nit: So if this test and the below tests pass, we can assume the wrapper exists from the mount

File: tasks/frontEnd/src/components/shared/AppActivityLoader/AppActivityLoader.test.ts [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+describe('AppActivityLoader.vue', () => {
+  it('renders correctly and matches snapshot when loading is false', () => {
+    const wrapper = mount(AppActivityLoader, {
+      props: { loading: false },
+      slots: {
+        default: '<p>Some default content here</p>',
+      },
+    });
+
+    // Check if the component is rendered
```
Comment (MarcusAl) [inline tasks/frontEnd/src/components/shared/AppActivityLoader/AppActivityLoader.test.ts:15]: Nit: Code comments should usually be describing why, not what something does

File: tasks/frontEnd/.eslintrc.cjs (hunk 2 above)
Comment (samcyn) [reply tasks/frontEnd/.eslintrc.cjs:23]: Yes this is inline with the Airbnb JavaScript Style Guide which I like to follow. However some of the other reasons I like it so is to 

1. Preventing Automatic Semicolon Insertion Issues
2. Improved Readability
3. Consistency
4. Tooling and Parsing
5. Collaboration

But as always this can be on/off depending on the team coding standard.

File: tasks/frontEnd/.github/workflows/pull-request-lint.yml (hunk 3 above)
Comment (samcyn) [reply tasks/frontEnd/.github/workflows/pull-request-lint.yml:1]: spot on I agree with this.

File: tasks/frontEnd/.github/workflows/pull-request-lint.yml [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+
+on: [push, pull_request]
```
[the remark this answers is not in the document]
Comment (samcyn) [reply tasks/frontEnd/.github/workflows/pull-request-lint.yml:3]: This is for demonstration purpose only. In a real production world, performing actions on every commit can indeed be resource-intensive and potentially costly, especially if using a CI service with usage-based billing. However, it worth noting this can help with

1. Early feedback
2. code quality
3. automation and enforcement.

Locally, each commits is linted without having to have any actions running anyway. So as mentioned, this is for demonstration purpose.

File: tasks/frontEnd/.husky/pre-commit [hunk 11]
```diff
@@ -0,0 +1 @@
+yarn lint
```
[the remark this answers is not in the document]
Comment (samcyn) [reply tasks/frontEnd/.husky/pre-commit:1]: yea it can be inline or here. It can be handy to fix eslint and formatting issues

File: tasks/frontEnd/e2e/auth-flow-login.spec.ts (hunk 5 above)
Comment (samcyn) [reply tasks/frontEnd/e2e/auth-flow-login.spec.ts:14]: As much as I would love to add as many scenarios as possible. I barely was out of time doing that. There are many more scenarios that can be considered including those you mentioned. I added a few of these for demonstrations and for the purpose of what can be done.

File: tasks/frontEnd/e2e/auth-flow-login.spec.ts (hunk 6 above)
Comment (samcyn) [reply tasks/frontEnd/e2e/auth-flow-login.spec.ts:48]: spot on! if you ask me if I have any improvement to add this is definitely one of them.

File: tasks/frontEnd/src/components/shared/AppCard/AppReviewCard.vue [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+}>();
+
+const props = withDefaults(defineProps<ReviewModel>(), {});
+
+/*notice api ratings are greater than 5, 
+and they are strings instead of numbers
+maybe not be needed in other scenario */
+const rating = computed(() => {
+  const value = props.review.rating;
+  const newValue = typeof value === 'string' ? parseInt(value) : value;
```
Comment (MarcusAl) [inline tasks/frontEnd/src/components/shared/AppCard/AppReviewCard.vue:25]: Should this type comparisons be needed if we're using `typescript`?

File: tasks/frontEnd/src/components/shared/AppCard/AppReviewCard.vue [hunk 13]
```diff
@@ -0,0 +1,102 @@
+<script setup lang="ts">
```
Comment (MarcusAl) [inline tasks/frontEnd/src/components/shared/AppCard/AppReviewCard.vue:1]: Can I ask your reasons for using typescript? Vue seems to have some type methods out of the box

Comment (MarcusAl) [reply tasks/frontEnd/src/components/shared/AppCard/AppReviewCard.vue:1]: It's great to see the usage though!

File: tasks/frontEnd/src/components/shared/AppPagination/AppPagination.test.ts [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+import { describe, it, expect, vi } from 'vitest';
+import AppPagination from '@/components/shared/AppPagination/index.vue';
+import { nextTick } from 'vue';
+
+describe('AppPagination', () => {
+
+  it('renders pagination correctly with page numbers and navigation buttons', async () => {
+    const wrapper = mount(AppPagination, {
+      props: {
+        totalCount: 15,
```
Comment (MarcusAl) [inline tasks/frontEnd/src/components/shared/AppPagination/AppPagination.test.ts:11]: Could we not set the wrapper once, and only change `totalCount` and `currentPage` across the tests, as those are the variables that change?

File: tasks/frontEnd/src/components/shared/AppPagination/index.vue [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+);
+
+const emit = defineEmits<{
+  (event: 'onPageChange', num: number): void;
+}>();
+
+const { paginationRange } = usePagination(props);
+
+const showPagination = computed(() => {
+  // If there are less than 2 times in pagination range we shall not render the component
```
Comment (MarcusAl) [inline tasks/frontEnd/src/components/shared/AppPagination/index.vue:30]: Isn't the usual practice to still show pagination, but disable the inputs when there are no extra pages to view?

File: tasks/frontEnd/src/components/shared/AppModal/index.vue [hunk 16]
```diff
@@ -0,0 +1,79 @@
+<script setup lang="ts">
```
Comment (MarcusAl) [inline tasks/frontEnd/src/components/shared/AppModal/index.vue:1]: Nice component here! The only issue I would have (and in other components) is the use of `index.vue` for naming. When grepping for files, it can make it card to spot the difference, and we should be more descriptive where possible

File: tasks/frontEnd/src/components/shared/AppActivityLoader/AppActivityLoader.test.ts [hunk 17]
```diff
@@ -0,0 +1,54 @@
+import { mount } from '@vue/test-utils';
+import { describe, it, expect } from 'vitest';
+import AppActivityLoader from '@/components/shared/AppActivityLoader/index.vue';
+import AppLoader from '@/components/shared/AppLoader/index.vue';
+
+describe('AppActivityLoader.vue', () => {
```
[the remark this answers is not in the document]
Comment (samcyn) [reply tasks/frontEnd/src/components/shared/AppActivityLoader/AppActivityLoader.test.ts:6]: I use typescript, it's a typescript project. I see no reason not to use it here.

File: tasks/frontEnd/src/composables/useAuth.ts [hunk 18]
```diff
@@ -0,0 +1,205 @@
+import { onMounted, onUnmounted, ref } from 'vue';
+import {
```
Comment (MarcusAl) [inline tasks/frontEnd/src/composables/useAuth.ts:2]: Any reason for not using `pinia` stores?

File: tasks/frontEnd/src/components/shared/AppActivityLoader/AppActivityLoader.test.ts (hunk 9 above)
Comment (samcyn) [reply tasks/frontEnd/src/components/shared/AppActivityLoader/AppActivityLoader.test.ts:15]: Yes so true. I added that for the purpose of illustrations/educating. If comments are added to educate, providing both the "what" and the "why" can be extremely valuable, especially in a learning context or for onboarding new team members (which was mentioned to me)

File: tasks/frontEnd/src/composables/useAuth.ts [hunk 19]
```diff
@@ -0,0 +1,205 @@
...
+
+  async function onRequestPassword(email: string) {
+    loading.value = true;
+    try {
+      await sendPasswordResetEmail(auth, email, {
+        url: `http://localhost:5173/login/?email=${email}`,
+        handleCodeInApp: true
+      });
+      loading.value = false;
+      router.replace({
```
Comment (MarcusAl) [inline tasks/frontEnd/src/composables/useAuth.ts:186]: Any reason for not using `push` instead of replace?

File: tasks/frontEnd/src/composables/useAuth.ts [hunk 20]
```diff
@@ -0,0 +1,205 @@
...
+  };
+};
+
+
+/**
+ * Provides custom logic to authenticate using specified strategy.
+ * @param strategy The authentication strategy ('google' or 'github').
+ * @returns An object containing loading state and authentication function.
+ */
+export const useProviderAuth = (strategy: 'google' | 'github' = 'github') => {
```
Comment (MarcusAl) [inline tasks/frontEnd/src/composables/useAuth.ts:121]: This is cool 👍

File: tasks/frontEnd/src/router/index.ts [hunk 21]
```diff
@@ -0,0 +1,79 @@
...
+    },
+    {
+      path: '/forgot-password',
+      name: 'forgotpassword',
+      component: () => import('@/pages/ForgotPasswordPage.vue')
+    }
+  ]
+});
+
+router.beforeEach(async (to, _, next) => {
```
Comment (MarcusAl) [inline tasks/frontEnd/src/router/index.ts:59]: 💯

File: tasks/frontEnd/src/components/shared/AppCard/AppReviewCard.vue (hunk 12 above)
Comment (samcyn) [reply tasks/frontEnd/src/components/shared/AppCard/AppReviewCard.vue:25]: if you noticed I added extra comment at the top. There are scenarios where type assertions or conversions may still be needed, especially if the type information is not fully accurate or if you're dealing with data from external sources (like APIs) where the type may not be as expected. This is one such scenario, typescript would be just fine taking what the 'value' for what it is at compile time. However, when the card is loaded, it's a different thing type of 'value' passed into it. I noticed that and did the comparison as approprite.

File: tasks/frontEnd/src/components/shared/AppCard/AppReviewCard.vue (hunk 13 above)
Comment (samcyn) [reply tasks/frontEnd/src/components/shared/AppCard/AppReviewCard.vue:1]: Like I mentioned I use typescript for virtually everything I do for it advantages. Yes out of the box vue exposes most of it apis but using typescript + vue is a match made in heaven. As you can see I can model what to the component emit easily. and define the props the component expect

File: tasks/frontEnd/src/components/shared/AppPagination/AppPagination.test.ts (hunk 14 above)
Comment (samcyn) [reply tasks/frontEnd/src/components/shared/AppPagination/AppPagination.test.ts:11]: Yes that can be refactored to do just that but note that the props are different per test case and once the component has mounted, we'll have to remount (rerender) the component in all the test case. This can be improve not doubt and I think the best way doing that is using a factory function instead.

File: tasks/frontEnd/src/components/shared/AppPagination/index.vue (hunk 15 above)
Comment (samcyn) [reply tasks/frontEnd/src/components/shared/AppPagination/index.vue:30]: yes it is. And as you noticed I put comment there. in some ui, pagination won't be shown if there are no pages aside the VERY first one. I disable a part of the logic if you notice, and this is so because I don't want the impression that pagination isn't add mere by looking at the UI.

File: tasks/frontEnd/src/components/shared/AppPagination/index.vue [hunk 22]
```diff
@@ -0,0 +1,91 @@
...
+	px-5 md:px-3 py-5 md:py-1.5 text-sm text-primary 
+	transition-all duration-300 hover:bg-neutral-100 
+`;
+</script>
+<template>
+  <nav v-show="showPagination" aria-label="pagination">
+    <ul class="list-style-none flex flex-wrap">
+      <li @click="onPrevious">
+        <button
+          :class="[anchorClass, `${currentPage === 1 ? 'cursor-not-allowed' : ''}`]"
```
[the remark this answers is not in the document]
Comment (samcyn) [reply tasks/frontEnd/src/components/shared/AppPagination/index.vue:58]: of course we all like it brief but have you consider ```currentPage === 1 && 'cursor-not-allowed'``` can actually evaluate to undefined. I don't want the possibility of having undefined in the class names

File: tasks/frontEnd/src/components/shared/AppModal/index.vue (hunk 16 above)
Comment (samcyn) [reply tasks/frontEnd/src/components/shared/AppModal/index.vue:1]: if you notice the pattern, I namespace the components with 'App'. Now for AppModal for instance, as a folder, I want a modal component itself, a test, a storybook story possibly. and some other files that may be use for the modal alone e.g the model of the props depending on how customizable I want it to be. Now if I name the component AppModal.vue inside the AppModal folder my import will look like this ```import AppModal from '../../AppModal/AppModal.vue'``` (this doesn't look nice). let say I name it Modal.vue(which doesn't sit with the rule off earlier) my import will be ```import AppModal from '../../AppModal/Modal.vue' ```( doesn't look pretty too). The advantage of having to use my approach is simple, you won't have to worry about the import statement look even in muti nested folders. Components and it required files can be grouped easily . Take note the component naming is still a multiword as convention required because inside defineOptions the name of the component is defined. it's a great approach but as always it's more about team coding style and convention settled for.

File: tasks/frontEnd/src/components/shared/AppTextAreaField/__test__/AppTextFieldWithHook.test.ts [hunk 23]
```diff
@@ -0,0 +1,65 @@
...
+    // expect(errorMessage.text()).toBe('Invalid email');
+
+    // // Correct the email and blur again
+    // await input.setValue('test@example.com');
+    // await input.trigger('blur');
+
+    // Check if the validation error message is removed
+    // expect(errorMessage.exists()).toBe(false);
+
+    expect(wrapper.html()).toMatchSnapshot();
```
[the remark this answers is not in the document]
Comment (samcyn) [reply tasks/frontEnd/src/components/shared/AppTextAreaField/__test__/AppTextFieldWithHook.test.ts:63]: well that can be true but not every time especially when you're worried about Visual Regression. AppTextAreaField is a component used in multple places and it's a very important component for the App so you will be interested to know if any visual changes happen to it. Same way stories in storybook are monitored for visual changes, snapshot test should be encourage for components that hold weight within the app, not only for visual testing but functionality as well.

File: tasks/frontEnd/src/composables/useAuth.ts (hunk 18 above)
Comment (samcyn) [reply tasks/frontEnd/src/composables/useAuth.ts:2]: Good question! Pinia isn't use because the auth state is maintain by vuefire store efficiently. If I wasn't using vuefire for example, I may consider Pinia or probably go with Provider/Inject api for an app of this size.

File: tasks/frontEnd/src/composables/useAuth.ts (hunk 19 above)
Comment (samcyn) [reply tasks/frontEnd/src/composables/useAuth.ts:186]: Good question. The reason is to prevent back navigation. Unlike push, replace replaces the current entry in the history stack with the new route. This means that users cannot navigate back to the login page using the browser's back button. Whichever use is always fine depending on the user experience wanted, level of security concern and taste.
